### PR TITLE
CFY-3717 add file based userstore

### PR DESCRIPTION
**Pull request by adamlavie**

### CODE REVIEW

File: flask_securest/authentication_providers/abstract_authentication_provider.py [hunk 1]
```diff
@@ -24,9 +32,35 @@ class AbstractAuthenticationProvider(object):
...
     """
-
     __metaclass__ = abc.ABCMeta
 
     @abc.abstractmethod
     def authenticate(self, userstore=None):
         raise NotImplementedError
+
+    @staticmethod
+    def retrieve_credentials_from_request():
```
Comment (noakup) [inline flask_securest/authentication_providers/abstract_authentication_provider.py]: this method should be implemented directly in the inheriting classes and not here. The headers are not generic are are part of the inner logic of each authentication provider. for example, a new authentication provider password2 could look for a header named pasword2, but other authentication providers should not be aware of this header at all. The logic you copied to the abstract is only relevant to the password authentication provider, and should not be available to other classes. please revert.

File: flask_securest/authorization_providers/role_loaders/simple_role_loader.py [hunk 2]
```diff
@@ -23,8 +23,11 @@
 class SimpleRoleLoader(AbstractRoleLoader):
 
     def get_roles(self):
+        # The simple role loader is designed to work with the file-userstore.
```
Comment (noakup) [inline flask_securest/authorization_providers/role_loaders/simple_role_loader.py]: as a matter of fact it is designed to work with any userstore that implements `get_roles` so I think we should just remove the comment on line 26

File: flask_securest/userstores/file_userstore.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+from watchdog.observers import Observer
+from watchdog.events import FileSystemEventHandler
+
+from flask_securest.userstores import simple
+
+
+class FileUserstore(simple.SimpleUserstore, FileSystemEventHandler):
+
+    def __init__(self, target_path):
+        self.target_path = target_path
```
Comment (noakup) [inline flask_securest/userstores/file_userstore.py]: maybe rename to userstore_path if that's indeed what it is? would be clearer

File: flask_securest/userstores/file_userstore.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+#  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  * See the License for the specific language governing permissions and
+#  * limitations under the License.
+import os
+import yaml
+
+from watchdog.observers import Observer
+from watchdog.events import FileSystemEventHandler
+
+from flask_securest.userstores import simple
```
Comment (noakup) [inline flask_securest/userstores/file_userstore.py:22]: I think it's more elegant to do:
`from flask_securest.userstores.simple import SimpleUserstore`
and then in your class definition inherit from `SimpleUserstore` and not `simple.SimpleUserstore`

File: flask_securest/authentication_providers/abstract_authentication_provider.py (hunk 1 above)
Comment (adamlavie) [inline flask_securest/authentication_providers/abstract_authentication_provider.py]: rename and move to utils. This method should called get_basic_http_authentication_info

File: flask_securest/authorization_providers/role_loaders/simple_role_loader.py [hunk 5]
```diff
@@ -23,8 +23,10 @@
 class SimpleRoleLoader(AbstractRoleLoader):
 
     def get_roles(self):
+        # todo: userstore should change to a mapping object that maps group/
```
Comment (noakup) [inline flask_securest/authorization_providers/role_loaders/simple_role_loader.py]: ??

File: flask_securest/authorization_providers/role_loaders/simple_role_loader.py [hunk 6]
```diff
@@ -23,8 +23,10 @@
...
+        # roles to cloudify roles
         userstore = current_app.securest_userstore_driver
-        principals = rest_security.get_principals_list() or {}
+        principals = rest_security.get_principals_list() or []
```
Comment (noakup) [inline flask_securest/authorization_providers/role_loaders/simple_role_loader.py:27]: :octopus: thanks for that!

**Pull request merged**